cmd/terminal.go

```diff
@@ -235,11 +296,117 @@ func (c *TerminalCmd) stopShell(conn *websocket.Conn) error {
 	return nil
 }
 
+func (c *TerminalCmd) record() {
+	f, err := os.Create(c.recordFile)
+	if err != nil {
+		log.Err(fmt.Sprintf("Can't create recording file: %s: %s", c.recordFile, err.Error()))
+	}
+	defer f.Close()
+
+	fz := gzip.NewWriter(f)
+	defer fz.Close()
+
+	data := TerminalRecordingHeader{
+		Version: terminalRecordingVersion,
+	}
+	copy(data.DeviceID[:], []byte(c.deviceID))
+	copy(data.DateTime[:], time.Now().String())
+	err = binary.Write(fz, binary.LittleEndian, data)
+	if err != nil {
+		log.Err(fmt.Sprintf("Header write failed: %s", err.Error()))
+	}
+	err = fz.Flush()
+	if err != nil {
+		log.Err(fmt.Sprintf("Header flush failed: %s", err.Error()))
+	}
+
+	log.Info(fmt.Sprintf("Recording to file: %s", c.recordFile))
+
+	e := gob.NewEncoder(fz)
+	for {
+		select {
+		case <-c.stopRecording:
+			return
+		case terminalOutput := <-c.terminalOutputChan:
+			o := TerminalRecordingData{
+				Type: terminalRecordingOutput,
+				Data: terminalOutput,
+			}
+			err = e.Encode(o)
+			fz.Flush()
+			if err != nil {
+				log.Err(fmt.Sprintf("Error encoding %q: %s", string(terminalOutput), err.Error()))
+				return
+			}
+		}
+	}
+}
+
+func (c *TerminalCmd) playback(w io.Writer) error {
+	f, err := os.Open(c.playbackFile)
+	if err != nil {
+		log.Err(fmt.Sprintf("Can't open %s: %s", c.playbackFile, err.Error()))
+		return err
+	}
+	defer f.Close()
+
+	fz, err := gzip.NewReader(f)
+	if err != nil {
+		return err
+	}
+	defer fz.Close()
+
+	var header TerminalRecordingHeader
+	err = binary.Read(fz, binary.LittleEndian, &header)
+	if err != nil {
+		log.Err(fmt.Sprintf("Can't read header: %s", err.Error()))
+		return err
+	}
+
+	log.Info(fmt.Sprintf("Playing back from file: %s", c.playbackFile))
+	log.Info(fmt.Sprintf("Device ID: %s", string(header.DeviceID[:])))
+	log.Info(fmt.Sprintf("Timestamp: %s", string(header.DateTime[:])))
+	log.Info("")
+
+	d := gob.NewDecoder(fz)
+	for {
+		var o TerminalRecordingData
+		err = d.Decode(&o)
+		if err != nil {
+			if err != io.EOF {
+				log.Err(fmt.Sprintf("Decoding error: %s", err.Error()))
+				return err
+			}
+			break
+		}
+		if o.Type == terminalRecordingOutput {
+			w.Write(o.Data)
+		}
+		time.Sleep(playbackSleep)
+	}
+	log.Info("\r")
```

Comment:
this is to make the carriage return, and the prompt to be presented nicely after playback ends and mender-cli returns?

Reply:
Yes, exactly. Not very important but I think harmless and nice to have.